File: backend/app/models/observatory.py
"""Observatory v2 API models — shapes EXACTLY per docs/build/api_contract.md (§1,§3,§4,§6).

Non-finite floats (NaN/Inf) serialize as null (§8). We use a field_serializer on every
optional-float field plus a model-level validator so a stray NaN never reaches the wire.
"""
from __future__ import annotations
from typing import Optional, Literal
import math
from pydantic import BaseModel, field_serializer, Field


def _clean_float(v):
    if v is None:
        return None
    try:
        f = float(v)
    except (TypeError, ValueError):
        return None
    return f if math.isfinite(f) else None


# ── §1 Dataset ─────────────────────────────────────────────────────────────────────

class DateRange(BaseModel):
    start: Optional[str] = None
    end: Optional[str] = None


class ColumnMap(BaseModel):
    timestamp: Optional[str] = None
    close: Optional[str] = None
    open: Optional[str] = None
    high: Optional[str] = None
    low: Optional[str] = None
    volume: Optional[str] = None


class Construction(BaseModel):
    beta_mode: Literal["none", "definitional", "frozen-ols", "rolling-INADMISSIBLE"] = "none"
    roll_masked: Optional[bool] = None  # null if N/A


class Dataset(BaseModel):
    dataset_id: str
    name: str
    source_file: str
    frequency: Literal["daily", "intraday", "unknown"]
    row_count: int
    date_range: DateRange
    columns: list[str]
    column_map: ColumnMap
    date_format: Literal["auto", "unix", "iso", "dd-mm-yyyy"] = "auto"
    timezone: Optional[str] = None
    is_spread: bool = False
    construction: Construction = Construction()
    created_at: str


# ── §3 QualityReport ────────────────────────────────────────────────────────────────

class Gap(BaseModel):
    from_: str = Field(alias="from", serialization_alias="from")
    to: str
    missing_bars: int

    model_config = {"populate_by_name": True}


class NonPositiveExample(BaseModel):
    time: str
    close: Optional[float] = None

    @field_serializer("close")
    def _s(self, v):
        return _clean_float(v)


class ExcisionWindow(BaseModel):
    from_: str = Field(alias="from", serialization_alias="from")
    to: str

    model_config = {"populate_by_name": True}


class NonPositivePrices(BaseModel):
    count: int = 0
    examples: list[NonPositiveExample] = []
    suggested_excision: Optional[ExcisionWindow] = None


class BackAdjustmentSeam(BaseModel):
    time: str
    jump: Optional[float] = None

    @field_serializer("jump")
    def _s(self, v):
        return _clean_float(v)


class QualityReport(BaseModel):
    row_count: int
    date_range: DateRange
    frequency: Literal["daily", "intraday", "unknown"]
    median_delta_days: Optional[float] = None
    gaps: list[Gap] = []
    n_gaps: int = 0
    duplicate_timestamps: int = 0
    non_positive_prices: NonPositivePrices = NonPositivePrices()
    nan_rows_dropped: int = 0
    back_adjustment_seams: list[BackAdjustmentSeam] = []
    warnings: list[str] = []

    @field_serializer("median_delta_days")
    def _s(self, v):
        return _clean_float(v)


# ── §6 Provenance ───────────────────────────────────────────────────────────────────

class Provenance(BaseModel):
    dataset_id: str
    dataset_hash: str
    as_of: Optional[str] = None
    params: dict
    mode: Literal["research", "verification"] = "research"
    prereg_id: Optional[str] = None
    exploratory_watermark: bool = False
    engine: str
    engine_version: str
    computed_at: str


# ── §2 ingestion response & series ──────────────────────────────────────────────────

class IngestResponse(BaseModel):
    dataset: Dataset
    quality: QualityReport


class DatasetList(BaseModel):
    datasets: list[Dataset]


class Bar(BaseModel):
    time: str
    open: Optional[float] = None
    high: Optional[float] = None
    low: Optional[float] = None
    close: Optional[float] = None
    volume: Optional[float] = None

    @field_serializer("open", "high", "low", "close", "volume")
    def _s(self, v):
        return _clean_float(v)


class SeriesResponse(BaseModel):
    dataset_id: str
    as_of: Optional[str] = None
    bars: list[Bar] = []
    forward_bars: list[Bar] = []


# ── §4.1 equilibrium ────────────────────────────────────────────────────────────────

class EquilibriumParams(BaseModel):
    snr: float = 1e-8
    kappa: float = 0.05
    warmup: int = 60


class EquilibriumRequest(BaseModel):
    dataset_id: str
    as_of: Optional[str] = None
    start: Optional[str] = None
    end: Optional[str] = None
    params: Optional[EquilibriumParams] = None
    mode: Literal["research", "verification"] = "research"
    prereg_id: Optional[str] = None


class EquilibriumRow(BaseModel):
    time: str
    close: Optional[float] = None
    mu_star: Optional[float] = None
    velocity: Optional[float] = None
    innovation: Optional[float] = None
    z: Optional[float] = None  # M1 — causal expanding-σ z; null before warmup σ defined
    gain: Optional[float] = None
    state_var: Optional[float] = None

    @field_serializer("close", "mu_star", "velocity", "innovation", "z", "gain", "state_var")
    def _s(self, v):
        return _clean_float(v)


class EquilibriumResponse(BaseModel):
    dataset_id: str
    as_of: Optional[str] = None
    params: EquilibriumParams
    series: list[EquilibriumRow] = []
    z_sigma_basis: str = "causal_expanding_innovation_std"
    provenance: Provenance


# ── §4.2 habitat ────────────────────────────────────────────────────────────────────

class HabitatWindow(BaseModel):
    start: str
    end: str


class HabitatParams(BaseModel):
    vr_qs: list[int] = [5, 10, 20]
    ns_null: int = 2000
    seed: int = 20260606


class HabitatRequest(BaseModel):
    dataset_id: str
    window: HabitatWindow
    as_of: Optional[str] = None
    deseason: bool = False
    params: Optional[HabitatParams] = None
    mode: Literal["research", "verification"] = "research"
    prereg_id: Optional[str] = None


class VRPoint(BaseModel):
    q: int
    vr: Optional[float] = None

    @field_serializer("vr")
    def _s(self, v):
        return _clean_float(v)


class SurrogateDistribution(BaseModel):
    null_min_vr: list[float] = []
    n: int = 0
    p10: Optional[float] = None
    p50: Optional[float] = None
    p90: Optional[float] = None
    frac_ge_real: Optional[float] = None  # == score/100

    @field_serializer("p10", "p50", "p90", "frac_ge_real")
    def _s(self, v):
        return _clean_float(v)


class CalibrationBadge(BaseModel):
    ou: float = 71.3
    rw: float = 49.2
    trend: float = 17.2
    status: str = "validated_non_inverting"


class RawVsDeseason(BaseModel):
    raw_score: Optional[float] = None
    deseason_score: Optional[float] = None
    verdict_changed: Optional[bool] = None

    @field_serializer("raw_score", "deseason_score")
    def _s(self, v):
        return _clean_float(v)


class HabitatResponse(BaseModel):
    dataset_id: str
    window: HabitatWindow
    deseason: bool = False
    score: Optional[float] = None
    real_min_vr: Optional[float] = None
    vr_curve: list[VRPoint] = []
    surrogate_distribution: SurrogateDistribution
    calibration_badge: CalibrationBadge = CalibrationBadge()
    raw_vs_deseason: Optional[RawVsDeseason] = None
    data_warning: Optional[str] = None
    provenance: Provenance

    @field_serializer("score", "real_min_vr")
    def _s(self, v):
        return _clean_float(v)

File: frontend/src/lib/observatory.ts
/**
 * Observatory v2 API client — typed against docs/build/api_contract.md (FROZEN).
 *
 * Thin client. NO statistic is computed here. Every number (μ*, z, VR, habitat score,
 * percentiles, frac_ge_real) comes from the v2 endpoints verbatim. The only array work
 * permitted on the frontend is BINNING null_min_vr[] into histogram bars for rendering
 * (display only) — never deriving a scalar (contract M5).
 *
 * Additive & isolated: this module does NOT touch src/lib/api.ts (the legacy /api/v1 client).
 */

const BASE = process.env.NEXT_PUBLIC_API_URL ?? 'http://localhost:8000';

// ── §1 Dataset ──────────────────────────────────────────────────────────────────────

export interface DateRange {
  start: string | null;
  end: string | null;
}

export interface ColumnMap {
  timestamp: string | null;
  close: string | null;
  open: string | null;
  high: string | null;
  low: string | null;
  volume: string | null;
}

export type BetaMode = 'none' | 'definitional' | 'frozen-ols' | 'rolling-INADMISSIBLE';

export interface Construction {
  beta_mode: BetaMode;
  roll_masked: boolean | null;
}

export type Frequency = 'daily' | 'intraday' | 'unknown';
export type DateFormat = 'auto' | 'unix' | 'iso' | 'dd-mm-yyyy';

export interface Dataset {
  dataset_id: string;
  name: string;
  source_file: string;
  frequency: Frequency;
  row_count: number;
  date_range: DateRange;
  columns: string[];
  column_map: ColumnMap;
  date_format: DateFormat;
  timezone: string | null;
  is_spread: boolean;
  construction: Construction;
  created_at: string;
}

// ── §3 QualityReport ──────────────────────────────────────────────────────────────────

export interface Gap {
  from: string;
  to: string;
  missing_bars: number;
}

export interface NonPositiveExample {
  time: string;
  close: number | null;
}

export interface ExcisionWindow {
  from: string;
  to: string;
}

export interface NonPositivePrices {
  count: number;
  examples: NonPositiveExample[];
  suggested_excision: ExcisionWindow | null;
}

export interface BackAdjustmentSeam {
  time: string;
  jump: number | null;
}

export interface QualityReport {
  row_count: number;
  date_range: DateRange;
  frequency: Frequency;
  median_delta_days: number | null;
  gaps: Gap[];
  n_gaps: number;
  duplicate_timestamps: number;
  non_positive_prices: NonPositivePrices;
  nan_rows_dropped: number;
  back_adjustment_seams: BackAdjustmentSeam[];
  warnings: string[];
}

// ── §6 Provenance ─────────────────────────────────────────────────────────────────────

export type Mode = 'research' | 'verification';

export interface Provenance {
  dataset_id: string;
  dataset_hash: string;
  as_of: string | null;
  params: Record<string, unknown>;
  mode: Mode;
  prereg_id: string | null;
  exploratory_watermark: boolean;
  engine: string;
  engine_version: string;
  computed_at: string;
}

// ── §2 ingestion & series ─────────────────────────────────────────────────────────────

export interface IngestResponse {
  dataset: Dataset;
  quality: QualityReport;
}

export interface DatasetList {
  datasets: Dataset[];
}

export interface Bar {
  time: string;
  open: number | null;
  high: number | null;
  low: number | null;
  close: number | null;
  volume: number | null;
}

export interface SeriesResponse {
  dataset_id: string;
  as_of: string | null;
  bars: Bar[];
  forward_bars: Bar[];
}

// ── §4.1 equilibrium ──────────────────────────────────────────────────────────────────

export interface EquilibriumParams {
  snr: number;
  kappa: number;
  warmup: number;
}

export interface EquilibriumRequest {
  dataset_id: string;
  as_of?: string | null;
  start?: string | null;
  end?: string | null;
  params?: EquilibriumParams | null;
  mode?: Mode;
  prereg_id?: string | null;
}

export interface EquilibriumRow {
  time: string;
  close: number | null;
  mu_star: number | null;
  velocity: number | null;
  innovation: number | null;
  z: number | null;
  gain: number | null;
  state_var: number | null;
}

export interface EquilibriumResponse {
  dataset_id: string;
  as_of: string | null;
  params: EquilibriumParams;
  series: EquilibriumRow[];
  z_sigma_basis: string;
  provenance: Provenance;
}

// ── §4.2 habitat ──────────────────────────────────────────────────────────────────────

export interface HabitatWindow {
  start: string;
  end: string;
}

export interface HabitatParams {
  vr_qs: number[];
  ns_null: number;
  seed: number;
}

export interface HabitatRequest {
  dataset_id: string;
  window: HabitatWindow;
  as_of?: string | null;
  deseason?: boolean;
  params?: HabitatParams | null;
  mode?: Mode;
  prereg_id?: string | null;
}

export interface VRPoint {
  q: number;
  vr: number | null;
}

export interface SurrogateDistribution {
  null_min_vr: number[];
  n: number;
  p10: number | null;
  p50: number | null;
  p90: number | null;
  frac_ge_real: number | null;
}

export interface CalibrationBadge {
  ou: number;
  rw: number;
  trend: number;
  status: string;
}

export interface RawVsDeseason {
  raw_score: number | null;
  deseason_score: number | null;
  verdict_changed: boolean | null;
}

export interface HabitatResponse {
  dataset_id: string;
  window: HabitatWindow;
  deseason: boolean;
  score: number | null;
  real_min_vr: number | null;
  vr_curve: VRPoint[];
  surrogate_distribution: SurrogateDistribution;
  calibration_badge: CalibrationBadge;
  raw_vs_deseason: RawVsDeseason | null;
  data_warning: string | null;
  provenance: Provenance;
}

// ── transport ─────────────────────────────────────────────────────────────────────────

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const res = await fetch(`${BASE}${path}`, {
    headers: { 'Content-Type': 'application/json' },
    ...init,
  });
  if (!res.ok) {
    const err = await res.json().catch(() => ({ detail: res.statusText }));
    throw new Error((err as { detail?: string }).detail ?? 'Request failed');
  }
  return res.json() as Promise<T>;
}

export interface IngestMapping {
  name?: string | null;
  column_map?: Partial<ColumnMap>;
  date_format?: DateFormat;
  timezone?: string | null;
}

export const observatory = {
  // §2 — upload CSV + explicit column mapping
  createDataset: async (file: File, mapping: IngestMapping): Promise<IngestResponse> => {
    const form = new FormData();
    form.append('file', file);
    form.append('mapping', JSON.stringify(mapping));
    const res = await fetch(`${BASE}/api/v2/datasets`, { method: 'POST', body: form });
    if (!res.ok) {
      const err = await res.json().catch(() => ({ detail: res.statusText }));
      throw new Error((err as { detail?: string }).detail ?? 'Upload failed');
    }
    return res.json() as Promise<IngestResponse>;
  },

  listDatasets: () => request<DatasetList>('/api/v2/datasets'),

  getDataset: (id: string) => request<Dataset>(`/api/v2/datasets/${encodeURIComponent(id)}`),

  // §2 — parsed bars + forward (evaluation-only) overlay, causal to as_of
  getSeries: (id: string, opts?: { as_of?: string | null; start?: string | null; end?: string | null }) => {
    const p = new URLSearchParams();
    if (opts?.as_of) p.set('as_of', opts.as_of);
    if (opts?.start) p.set('start', opts.start);
    if (opts?.end) p.set('end', opts.end);
    const qs = p.toString();
    return request<SeriesResponse>(`/api/v2/datasets/${encodeURIComponent(id)}/series${qs ? `?${qs}` : ''}`);
  },

  getQuality: (id: string) => request<QualityReport>(`/api/v2/datasets/${encodeURIComponent(id)}/quality`),

  // §4.1 — Kalman μ* + causal z
  equilibrium: (body: EquilibriumRequest) =>
    request<EquilibriumResponse>('/api/v2/analysis/equilibrium', {
      method: 'POST',
      body: JSON.stringify(body),
    }),

  // §4.2 — surrogate-relative habitat score (cloud is mandatory in the response)
  habitat: (body: HabitatRequest) =>
    request<HabitatResponse>('/api/v2/analysis/habitat', {
      method: 'POST',
      body: JSON.stringify(body),
    }),
};

// ── display-only helper: bin null_min_vr[] into histogram BARS (NOT a scalar). ──────────
// Per M5 this is permitted for RENDERING only; it derives no displayed number.
export interface HistogramBin {
  x0: number;
  x1: number;
  count: number;
}

export function binForHistogram(values: number[], nbins = 24): HistogramBin[] {
  const finite = values.filter((v) => Number.isFinite(v));
  if (finite.length === 0) return [];
  let lo = Math.min(...finite);
  let hi = Math.max(...finite);
  if (lo === hi) {
    hi = lo + 1e-9;
  }
  const width = (hi - lo) / nbins;
  const bins: HistogramBin[] = Array.from({ length: nbins }, (_, i) => ({
    x0: lo + i * width,
    x1: lo + (i + 1) * width,
    count: 0,
  }));
  for (const v of finite) {
    let idx = Math.floor((v - lo) / width);
    if (idx >= nbins) idx = nbins - 1;
    if (idx < 0) idx = 0;
    bins[idx].count += 1;
  }
  return bins;
}
